book/visual-effects.md

```diff
@@ -5,312 +5,396 @@ prev: security
                     browser.handle_down()
-            if event.type == sdl2.SDL_TEXTINPUT:
+            elif event.type == sdl2.SDL_TEXTINPUT:
                 browser.handle_key(event.text.text.decode('utf8'))
-            if event.type == sdl2.SDL_QUIT:
-                running = False
-                break
-
-    sdl2.SDL_DestroyWindow(sdl_window)
-    sdl2.SDL_Quit()
 ```
 
-Next factor a bunch of the tasks of drawing into helper methods. The
-implementation in Skia should be relatively self-explanatory (there is also
-more complete documentation[here](https://kyamagu.github.io/skia-python/) or at
-the [Skia site](https://skia.org)):
+You can now remove all of the `bind` calls in the `Browser`
+constructor; this main loop replaces them. Also note that I've changed
+the signatures of the various `handle_xxx` methods; you'll need to
+make analogous changes in `Browser` where they are defined.
+
+Now our browser is creating an SDL window can draw to it via Skia. But
+most of the browser codebase is still using Tkinter drawing commands,
+which we now need to replace. Skia is a bit more verbose than Tkinter,
+so let's abstract over some details with helper functions.[^skia-docs]
+
+[skia-docs]: Consult the [Skia][skia] and [Skia-Python][skia-python]
+documentation for more on the Skia API.
+
+To draw a line, you use Skia's `Path` object:
 
 ``` {.python}
-def draw_polyline(surface, x1, y1, x2, y2, x3=None,
-    y3=None, fill=False):
-    path = skia.Path()
-    path.moveTo(x1, y1)
-    path.lineTo(x2, y2)
-    if x3:
-        path.lineTo(x3, y3)
-    paint = skia.Paint()
-    paint.setColor(skia.ColorBLACK)
-    if fill:
-        paint.setStyle(skia.Paint.kFill_Style)
-    else:
-        paint.setStyle(skia.Paint.kStroke_Style)
+def draw_line(surface, x1, y1, x2, y2):
+    path = skia.Path().moveTo(x1, y1).lineTo(x2, y2)
+    paint = skia.Paint(Color=skia.ColorBLACK)
+    paint.setStyle(skia.Paint.kStroke_Style)
     paint.setStrokeWidth(1);
     with surface as canvas:
         canvas.drawPath(path, paint)
+```
 
-class Browser:
-    # ...
-    def draw_text(self, x, y, text, font, color=None):
-        paint = skia.Paint(
-            AntiAlias=True, Color=color_to_sk_color(color))
-        with self.skia_surface as canvas:
-            canvas.drawString(
-                text, x, y - font.getMetrics().fAscent,
-                font, paint)
-
-    def draw_rect(self, rect, fill=None, width=1):
-        paint = skia.Paint()
-        if fill:
-            paint.setStrokeWidth(width);
-            paint.setColor(color_to_sk_color(fill))
-        else:
-            paint.setStyle(skia.Paint.kStroke_Style)
-            paint.setStrokeWidth(1);
-            paint.setColor(skia.ColorBLACK)
-        with self.skia_surface as canvas:
-            canvas.drawRect(rect, paint)
+To draw text, you use `drawString`:
+
+``` {.python}
+def draw_text(surface, x, y, text, font, color=None):
+    sk_color = color_to_sk_color(color)
```

Comment:
Looks like I forgot to show the code for color_to_sk_color...

Reply:
Good point, I'll add it.